TKH-3142 Make the list of service accounts additional object writable as list of uuids

haster opened the pull request:
- findByUuid logic now also works if the linkable is within a linkablewrapper
- To avoid only ever returning uuids, findByUuid logic no doesn't work in a readonly context (meaning a field that is either read-only, or it's parent field is read-only, etc.)


- This enables the linking of a set of service-accounts when creating a group-on-system by specifying their uuids
- This also means that a number of `field_uuid` fields that were effectively read-only, are now just `field` and return the entire object. See `parent` in (client_)vaultrecord.
- Such a list of uuids is treated as a Set, so TF won't balk over the ordering.

See https://github.com/topicuskeyhub/terraform-provider-keyhub/pull/92 for the total set of changes.
See https://github.com/topicusonderwijs/keyhub/pull/2320 for the TF test changes in KeyHub

haster committed:
TKH-3142 Generate findByUuid also for wrapped lists but only if not readonly

haster committed:
TKH-3142 Gather a map of types for readwrite but also readonly context

haster committed:
TKH-3142 Add and use SDKInterfaceTypeName where appropriate

The SDKTypeName methode really returned the interface type, not the concrete type.
So we renamed that method, including all usages, to
'SDKInterfaceTypeName' so the naming fits betteri.
We also need the actual type in a few spots in which we now akwardly recreate it based on the wrong property which doesn't fit as well.
So we added an actual 'SDKTypeName' method that returns the actual type
and use that.

haster committed:
TKH-3142 Store the 'inreadonlycontext' with the type for a property

We now store for a type whether or not it was generated for a read-only
property.
We use that property for the 'GoTypeName', which in turn gets used in
the template files.

haster committed:
TKH-3142 Add virtual findByUuid type for nested object in wrapper type

haster committed:
Revert "TKH-3142 Add virtual findByUuid type for nested object in wrapper type"

This reverts commit 6c76da6730d1ffb02bdf1ac17459f18eedfe6c7c.

haster committed:
TKH-3142 Render property for list of findByUuid types as '_uuid'

haster committed:
TKH-3142 Ensure every subtypes is created for either context

Otherwise we could have only half a subtype hierarchy in a certain context, which is generally not what we want.

papegaaij commented on main.go:
```diff
@@ -215,15 +215,37 @@ func merge(template string, suffix string, t *template.Template, model any) {
+			log.Printf("Keeping read/write type %s", readWriteContextType.GoTypeName())
+
+			if readOnlyContextType != nil {
+				// if reflect.TypeOf(readWriteContextType) != reflect.TypeOf(readOnlyContextType) {
+				ret = append(ret, readOnlyContextType)
+				i++
+				log.Printf("Also keeping readonly type %s", readOnlyContextType.GoTypeName())
+				// } else {
+				// 	log.Printf("Not keeping readonly type %s since it is the same", readOnlyContextType.GoTypeName())
+				// }
```
Deze comments kunnen nog even opgeruimd worden.

papegaaij reviewed (approved):
Het is complexe code, maar ik zie geen problemen. Enkel stukje commented code.

haster committed:
TKH-3142 Make sure to reset the $type parameter when casting to a primer

Otherwise we might send a Primer with the $type value of a full entity,
which could lead to validation errors when jackson tries to deserialize.

haster committed:
TKH-3142 Treat uuid list as set to avoid having to reorder

haster committed:
TKH-3142 Fix accidentally introduced stackoverflow

haster committed:
TKH-3142 Correctly handle empty set type for linkablewrapper.items

haster committed:
TKH-3142 Code and debug logging cleanup

papegaaij reviewed (approved):
Ziet er prima uit zo

haster committed:
Merge remote-tracking branch 'origin/main' into task-tkh-3142
